p2wsh and p2sh-p2wsh address in decodescript

fivepiece opened the pull request:
Attempts to address #12244 .  `p2wsh` addresses are returned only for scripts that are neither `p2sh` nor any witness program.

fivepiece committed:
p2wpkh, p2wsh and p2sh-nested scripts in decodescript

plus tests

fivepiece commented:
Again I'm only considering the edge cases after clicking "create pull request"..  What if the script contains a checksig with an uncompressed pubkey?
Obviously not ready.

MeshCollider commented:
@fivepiece feel free to add [WIP] to the title of your PR if its work in progress :) It's usually fine to open up a PR before its completely ready because other contributors might like to give feedback on the approach or high level concept ACK/NACKs regardless

fivepiece commented:
Cheers, title changed

fivepiece commented:
I believe this push does things more along the lines that you guys suggested.  A couple things :

1. If you set a label on an imported privkey, then try to set different labels to either its p2pkh or p2wpkh addresses, they will not show up in dumpwallet.  The label set to the privkey will show up instead.
2. If you set a label on an imported privkey, then try to set a different label on its p2sh-p2wpkh address, its label will be used in the umpwallet entry.
3. Change addresses, even if generated by `getrawchangeaddress` will not retain their original address type, that's because they have no label.

Here's an example dump :
https://gist.github.com/fivepiece/fe8e708264c390e0e943e54cec6d1162

You can see imported compressed and uncompressed keys and keys that were generated by `getnewaddress` retain their original type.
Things like `reserve` , `change` and `hdmaster` (and other non `label`s) will have their output address set to the user's global setting.
I've also added the case of outputting p2sh-p2wpkh addresses in the dump for non address book keys (in the `switch` statement) as the user might actually want to dump reserved keys as p2sh-p2wpkh.

In the gist, the line :
```
0014d6e506edcaf58910c27c4d8cbe94014552297808 1970-01-01T00:00:00Z script=1 # addr=2N1csGz3tHSBwXaLsHULSWj2PSphzoWvFPX
```
is a result of `importpubkey` with an compressed pubkey.  It's the p2sh-p2wpkh script for it.  I think this is new.  It doesn't happen for `importpubkey` with an uncompressed pubkey.

I'll go over the tests now.  It's getting an address entry that looks like `mx5AyoDPJViVFaAu9VFMQ1MbwiUtTq8qJg,2MxskE9rmHDyK6pwdwM55ZFdAk3HtkQiwNW` so it currently breaks.

Cheers

fivepiece commented:
Ideas welcome, returning a p2wsh for a p2pk or multisig with an uncompressed pubkey can result in fund loss, then again `decodescript` already returns p2sh addresses even for complete garbage.
What should this change to `decodescript` take into account when deciding whether to post a p2wsh address to the user?

fivepiece commented:
Agreed.  Keeping the simple check that this is not a bare witness program seems to be in line with the check above it not letting p2sh be re-encoded into p2sh.
This commit adds p2sh-p2wsh as well.

promag reviewed (commented):
Missing tests for new keys.

Doesn't make sense to move this to `ScriptPubKeyToUniv`?

fivepiece commented:
@promag wouldn't moving this to `ScriptPubKeyToUniv` cause script translation to all these types of addresses to fire on `decoderawtransaction` as well?
Sorry if I misunderstood your comment.

I will be looking into adding tests as well.

fivepiece commented:
I've made some changes to this:

1. Segwit stuff is handled in a generic way using `GetScriptForWitness()`.  This allows to return a more appropriate p2wpkh for p2pkh and p2pk scriptpubkeys, and (currently) p2wsh for all the rest.
2. The data returned for the segwit scripts is its own univalue object (@promag suggestion), so it includes more useful info like the segwit scriptpubkey and type, along with the bech32 address.

Various examples of the output can be seen here :
https://gist.github.com/fivepiece/e25b0fcdf8303fdbfd56363ae50dec15

Caveats :

1. For the segwit scriptpubkeys, we always have `"reqSigs": 1`.
2. It's still possible to get a segwit program address for scripts using checksigs with uncompressed pubkeys.  This is because `GetScriptForWitness()` doesn't check `vSolutions` for presense of such pubkeys.  It's probably be possible to change GetScriptForWitness to check for that when bare pubkeys are invloved in the script, but out of scope for this PR.

fivepiece commented:
This is pretty much a complete rewrite taking into account @sipa's suggestion (on irc) to use the same logic from `validateaddress` in here.

1. Uncompressed pubkeys are handled correctly - meaning standard scripts containing them will not be encoded in segwit scriptpubkeys.
2. Added a few tests on top of the existing `rpc_decodescript.py`.  I couldn't find a way to easily test that segwit output is **not** returned for standard scripts with uncompressed pubkeys.  Suggestions on how to do that are welcome.
3. Updated https://gist.github.com/fivepiece/e25b0fcdf8303fdbfd56363ae50dec15 with examples showing segwit output is disabled for uncompressed pubkey scripts.

(please restart the travis job if possible, it hasn't failed, just cancelled before it began)

instagibbs commented:
kicked travis.

> I couldn't find a way to easily test that segwit output is not returned for standard scripts with uncompressed pubkeys.

hm? Your gist has examples that you could use.

fivepiece commented:
Specifically, I don't know how to assert that a value (say `rpc_result['segwit']`) does not exist in the returned univalue.

instagibbs commented:
in python `'segwit' not in rpc_result` ?

fivepiece commented:
Ah, I was trying to bend `assert_array_result` to do the check but your way works.

fivepiece commented:
Added tests for checking `'segwit'` value is not returned when inappropriate.

instagibbs commented on test/functional/rpc_decodescript.py:
```diff
@@ -59,18 +59,24 @@ def decodescript_script_pub_key(self):
         rpc_result = self.nodes[0].decodescript('52' + push_public_key + push_public_key + push_public_key + '53ae')
         assert_equal('2 ' + public_key + ' ' + public_key + ' ' + public_key +  ' 3 OP_CHECKMULTISIG', rpc_result['asm'])
+        # multisig in P2WSH
+        assert_equal('0 0f320decca000221b611117448134c78dd220cd22384ac80025b59e43c8698e8', rpc_result['segwit']['asm'])
```
could you generate this dynamically instead of hardcoding the long hex?

instagibbs commented on test/functional/rpc_decodescript.py:
```diff
@@ -103,6 +109,8 @@ def decodescript_script_pub_key(self):
         rpc_result = self.nodes[0].decodescript('63' + push_public_key + 'ad670320a107b17568' + push_public_key + 'ac')
         assert_equal('OP_IF ' + public_key + ' OP_CHECKSIGVERIFY OP_ELSE 500000 OP_CHECKLOCKTIMEVERIFY OP_DROP OP_ENDIF ' + public_key + ' OP_CHECKSIG', rpc_result['asm'])
+        # CLTV script in P2WSH
+        assert_equal('0 01f79ceb012cbcefb251f7843b827a3e285f40ddc5014d35e2486716cbacf5f1', rpc_result['segwit']['asm'])
```
could you generate this dynamically instead of hardcoding the long hex?

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/12321/commits/785bd02cda896c92190a8a9bba3d1b40fd2812b9

I think you can squash almost all these commits.

fivepiece commented:
Cool, I'll squash it all into a single commit if this current change looks good, or two if it's better to separate work done on core from the tests.  Let me know which is better.

instagibbs commented:
whatever's easier to you, 1 or 2

fivepiece commented:
Two commits now.  No problem changing later on if needed.

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -563,6 +563,41 @@ UniValue decodescript(const JSONRPCRequest& request)
+            if ((which_type == TX_PUBKEY) || (which_type == TX_MULTISIG)) {
+                for (const auto& solution : solutions_data) {
+                    if ((solution.size() != 1) && !CPubKey(solution).IsCompressed()) {
+                        fFoundUncompressedPubkeys = true;
```
just return `r` here. then remove the conditional later?

fivepiece replied:
Ah I see what you mean, looking into this.

instagibbs reviewed (approved):
utACK

fivepiece commented:
Good call.  All in a single commit now.

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/12321/commits/4f933b3d23010d3b03998460290faed97cd6f236

fivepiece committed:
list the types of scripts we should consider for a witness program

fivepiece commented:
Updated to address @laanwj's comment

laanwj commented:
Thank you
utACK https://github.com/bitcoin/bitcoin/pull/12321/commits/41ff9675a92f6c080a4bb171e286060ef864a803
